server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -0,0 +1,411 @@
+        super.consumeResult(result, () -> {});
+        QuerySearchResult querySearchResult = result.queryResult();
+        pendingMerges.consume(querySearchResult, next);
+        progressListener.notifyQueryResult(querySearchResult.getShardIndex());
+    }
+
+    @Override
+    SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
+        if (pendingMerges.hasPendingMerges()) {
+            throw new AssertionError("partial reduce in-flight");
+        } else if (pendingMerges.hasFailure()) {
+            throw pendingMerges.getFailure();
+        }
+
+        logger.trace("aggs final reduction [{}] max [{}]", aggsCurrentBufferSize, aggsMaxBufferSize);
+        // ensure consistent ordering
+        pendingMerges.sortBuffer();
+        final TopDocsStats topDocsStats = pendingMerges.consumeTopDocsStats();
+        final List<TopDocs> topDocsList = pendingMerges.consumeTopDocs();
+        final List<InternalAggregations> aggsList = pendingMerges.consumeAggs();
+        SearchPhaseController.ReducedQueryPhase reducePhase = controller.reducedQueryPhase(results.asList(), aggsList,
+            topDocsList, topDocsStats, pendingMerges.numReducePhases, false, aggReduceContextBuilder, performFinalReduce);
+        progressListener.notifyFinalReduce(SearchProgressListener.buildSearchShards(results.asList()),
+            reducePhase.totalHits, reducePhase.aggregations, reducePhase.numReducePhases);
+        return reducePhase;
+    }
+
+    private MergeResult partialReduce(MergeTask task,
+                                      TopDocsStats topDocsStats,
+                                      MergeResult lastMerge,
+                                      int numReducePhases) {
+        final QuerySearchResult[] toConsume = task.consumeBuffer();
+        if (toConsume == null) {
+            // the task is cancelled
+            return null;
+        }
+        // ensure consistent ordering
+        Arrays.sort(toConsume, Comparator.comparingInt(QuerySearchResult::getShardIndex));
+
+        for (QuerySearchResult result : toConsume) {
+            topDocsStats.add(result.topDocs(), result.searchTimedOut(), result.terminatedEarly());
+        }
+
+        final TopDocs newTopDocs;
+        if (hasTopDocs) {
+            List<TopDocs> topDocsList = new ArrayList<>();
+            if (lastMerge != null) {
+                topDocsList.add(lastMerge.reducedTopDocs);
+            }
+            for (QuerySearchResult result : toConsume) {
+                TopDocsAndMaxScore topDocs = result.consumeTopDocs();
+                setShardIndex(topDocs.topDocs, result.getShardIndex());
+                topDocsList.add(topDocs.topDocs);
+            }
+            newTopDocs = mergeTopDocs(topDocsList,
+                // we have to merge here in the same way we collect on a shard
+                topNSize, 0);
+        } else {
+            newTopDocs = null;
+        }
+
+        final Serialized<InternalAggregations> newAggs;
```

Comment:
I think it is more clear to write this is `DelayableWriteable.Serialized<InternalAggregations>`.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/58461/commits/a6c4ff4545a5423f6f53f92e07fb35e3e2ff189a